src/Ethereum/Transaction.cpp

```diff
@@ -212,4 +217,179 @@ TransactionEip1559::buildERC1155Transfer(const uint256_t& nonce,
     return std::make_shared<TransactionEip1559>(nonce, maxInclusionFeePerGas, maxFeePerGas, gasPrice, tokenContract, 0, TransactionNonTyped::buildERC1155TransferFromCall(from, to, tokenId, value, data));
 }
 
-} // namespace TW::Ethereum
+/// TransactionEip4337
+Data TransactionEip4337::preHash(const uint256_t chainID) const {
+    auto params = ABI::ParamTuple(std::vector<std::shared_ptr<ABI::ParamBase>>{
+        std::make_shared<ABI::ParamByteArrayFix>(32, Hash::keccak256(serialize(chainID))),
+        std::make_shared<ABI::ParamAddress>(entryPoint),
+        std::make_shared<ABI::ParamUInt256>(chainID)
+    });
+    Data encoded;
+    params.encode(encoded);
+    const auto hash = Hash::keccak256(encoded);
+    const auto hashStr = std::string(hash.begin(), hash.end());
+    return MessageSigner::generateMessage(hashStr);
+}
+
+Data TransactionEip4337::serialize(const uint256_t chainID) const {
+    auto params = ABI::ParamTuple(std::vector<std::shared_ptr<ABI::ParamBase>>{
+        std::make_shared<ABI::ParamAddress>(sender),
+        std::make_shared<ABI::ParamUInt256>(nonce),
+        std::make_shared<ABI::ParamByteArray>(initCode),
+        std::make_shared<ABI::ParamByteArray>(payload),
+        std::make_shared<ABI::ParamUInt256>(gasLimit),
+        std::make_shared<ABI::ParamUInt256>(verificationGasLimit),
+        std::make_shared<ABI::ParamUInt256>(preVerificationGas),
+        std::make_shared<ABI::ParamUInt256>(maxFeePerGas),
+        std::make_shared<ABI::ParamUInt256>(maxInclusionFeePerGas),
+        std::make_shared<ABI::ParamByteArray>(paymasterAndData),
+        std::make_shared<ABI::ParamByteArray>(parse_hex("0x"))
+    });
+    Data serialized;
+    params.encode(serialized);
+
+    return Data(serialized.begin(), serialized.end() - 32); // remove trailing word (zero-length signature)
+}
+
+Data TransactionEip4337::encoded(const Signature& signature, const uint256_t chainID) const {
+    Data rawSignature = signature.full;
+    rawSignature[64] += 27;
```

Comment:
This signature doesn't need EIP155 repay attack protection? (let's have a function that get the full signature from r,s,v similar to my precedent comment)